Precompute and store attribute and executor id together in a single h…

baldersheim opened the pull request:
…ashmap.

the executor id so complex fields use the same executor.
@toregge and @geirst PR

Commit:
Precompute and store attribute and executor id together in a single hashmap.
the executor id so complex fields use the same executor.

toregge commented on searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp:
```diff
@@ -474,9 +474,15 @@ AttributeWriter::AttributeWriter(const proton::IAttributeManager::SP &mgr)
+    for (auto attr : getWritableAttributes()) {
+        vespalib::stringref name = attr->getName();
+        vespalib::stringref prefix = name.substr(0, name.find('.'));
+        _attrMap[attr->getName()] = AttrWithId(attr, _attributeFieldWriter.getExecutorId(prefix));
```
This doesn't match how executor id is set in FieldContext constructor.

toregge reviewed (changes requested):
Please ensure consistent mapping to executor id.

Commit:
Choose executorId the same way both places.

baldersheim commented:
Unified selection of executorId.

toregge commented:
Doesn't match executor id used in AttributeWriter::heartBeat(), AttributeWriter::onReplayDone(),
AttributeWriter::compactLidSpace(), FlushableAttribute::initFlush(), ExclusiveAttributeReadAccessor::takeGuard(), AttributeManager::asyncForEachAttribute(), 
FilterAttributeManager::asyncForEachAttribute(), GidToLidChangeListener::GidToLidChangeListener(),
and allocShrinker() (not necessarily a complete list).

Commit:
Add getNamePrefix to get the prefix before any '.'

Commit:
Rewrite test to enure AttributeWriter is given a fully populated AttributeManager

Commit:
- Remove the execute(string, ...) and force the use of ExecutorId.
- Remove some double bookkeeping in AttributeWriter.
- Ensure that we always use attribute.getNamePrefix() to compute executor id.

Commit:
Use correct prefix name when selecting executor.

baldersheim commented:
Now, I think I have caught all of them ..

Commit:
Add test for getNamePrefix

Commit:
Add missing ')'

toregge reviewed (approved):
LGTM

geirst commented on searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.h:
```diff
@@ -58,11 +58,15 @@ class AttributeWriter : public IAttributeWriter
+    AttrMap                   _attrMap;
 
     void setupWriteContexts();
+    void setupAttriuteMapping();
```
Typo: Attriute -> Attribute

geirst reviewed (commented):
@baldersheim FYI
